### PR TITLE
Forward data packets

### PR DESCRIPTION
**Pull request by boks1971**
Next stop on data tracks support. This is set up to merge into my previous [PR](https://github.com/livekit/livekit/pull/4089) branch. This adds support for forwarding data track packets.

Some bits TODO:
- data track packet tests
-  for some reason, the multi node scenario test does not receive all packets or receives packets after some unexplained delay, have to dig into this further. It is not consistent. So, has been a hard one to chase.

Will probably looking at getting this merged into the previous branch and into `master` to get minimal version of this feature out. Have to make cloud changes also to get this working across cloud. Will then iterate on it based on usage.

### CODE REVIEW

File: pkg/rtc/datatrack.go [hunk 1]
```diff
@@ -106,3 +125,29 @@ func (d *DataTrack) IsSubscriber(subID livekit.ParticipantID) bool {
 	_, ok := d.subscribedTracks[subID]
 	return ok
 }
+
+func (d *DataTrack) AddDataDownTrack(dts types.DataTrackSender) error {
+	if d.closed.IsBroken() {
+		return errReceiverClosed
+	}
```
Comment (milos-lk) [inline pkg/rtc/datatrack.go:132]: Since we have this check - just wanted to check if we have/need atomicity with the actions below in this function - would something bad happen if we check closed -  we see it's not broken and proceed but then immediately fuse gets broken?

Comment (boks1971) [reply pkg/rtc/datatrack.go:132]: Will think about it more.

It should be fine I think. The down track will get added, but when delete gets called, it should be cleaned up.

File: pkg/rtc/datatrack.go [hunk 2]
```diff
@@ -106,3 +125,29 @@ func (d *DataTrack) IsSubscriber(subID livekit.ParticipantID) bool {
...
+	return nil
+}
+
+func (d *DataTrack) DeleteDataDownTrack(subscriberID livekit.ParticipantID) {
+	d.downTrackSpreader.Free(subscriberID)
+	d.params.Logger.Infow("data downtrack deleted", "subscriberID", subscriberID)
+}
+
+func (d *DataTrack) HandlePacket(data []byte, packet *datatrack.Packet) {
+	d.params.Logger.Infow("received data track message", "id", d.ID(), "name", d.Name(), "size", len(data))
```
Comment (milos-lk) [inline pkg/rtc/datatrack.go]: just checking if you intentionally added info level logging on packet level?

Comment (boks1971) [reply pkg/rtc/datatrack.go]: yeah, it is intentional for initial debugging. But, will remove it before merging.

File: test/client/datatrack_writer.go [hunk 3]
```diff
@@ -14,18 +14,64 @@
...
+		default:
+			packets := datatrack.GenerateRawDataPackets(d.handle, seqNum, frameNum, 1, rand.Intn(2048)+1, 100*time.Millisecond)
+			for _, packet := range packets {
+				if err := d.transport.SendDataTrackMessage(packet); err != nil {
+					logger.Errorw("could not send data track packet", err)
+				}
+			}
+
+			var lastPacket datatrack.Packet
+			if err := lastPacket.Unmarshal(packets[len(packets)-1]); err == nil {
```
Comment (milos-lk) [inline test/client/datatrack_writer.go]: nit: I see `GenerateRawDataPackets` can return nil in some corner cases - maybe we should guard this to avoid `packets[len(packets)-1]` in that case?

Comment (boks1971) [reply test/client/datatrack_writer.go]: Good catch. Will do.

**Review by milos-lk — APPROVED** [review]

**Comment by boks1971** [comment]
Merging this to the other PR branch, thank you @milos-lk . If others do have comments, please add and I will address.
